### PR TITLE
Replace use of legacy identity custom model with identity backend

### PR DESCRIPTION
**Pull request by CoderHam**
- [x] Fixed trace in new Backend API - will fix failure in cmdline_trace and we can re-enable related test in L0_batcher
 - [x] Fixed failure in L0_lifecycle - test_model_reload_fail due to deps on non-GPU identity backend build.

### CODE REVIEW

File: Dockerfile.QA [hunk 1]
```diff
@@ -165,51 +165,51 @@ RUN mkdir -p qa/custom_models/custom_sequence_int32/1 && \
...
         qa/L0_client_timeout/. && \
-    cp /tmp/tritonbuild/tritonserver/build/test-util/install/lib/libidentity.so \
+    cp /tmp/tritonbuild/install/backends/identity/libtriton_identity.so \
         qa/L0_client_valgrind/. && \
-    cp /tmp/tritonbuild/tritonserver/build/test-util/install/lib/libidentity.so \
+    cp /tmp/tritonbuild/install/backends/identity/libtriton_identity.so \
         qa/L0_client_memory_growth/. && \
-    cp /tmp/tritonbuild/tritonserver/build/test-util/install/lib/libidentity.so \
+    cp /tmp/tritonbuild/install/backends/identity/libtriton_identity.so \
         qa/L0_custom_legacy/. && \
```
Comment (GuanLuo) [inline Dockerfile.QA:213]: You can probably drop this test, its name suggests that it is not needed after we deprecate legacy custom backend

File: qa/L0_client_memory_growth/models/custom_identity_int32/config.pbtxt [hunk 2]
```diff
@@ -25,10 +25,10 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 name: "custom_identity_int32"
-platform: "custom"
+platform: "identity"
 max_batch_size: 1024
 version_policy: { latest { num_versions: 1 }}
-default_model_filename: "libidentity.so"
+default_model_filename: "libtriton_identity.so"
```
Comment (GuanLuo) [inline qa/L0_client_memory_growth/models/custom_identity_int32/config.pbtxt:31]: Don't need this

File: qa/L0_client_memory_growth/test.sh [hunk 3]
```diff
@@ -54,7 +54,7 @@ SERVER_ARGS="--model-repository=$DATADIR"
 source ../common/util.sh
 
 mkdir -p $DATADIR/custom_identity_int32/1
-cp libidentity.so $DATADIR/custom_identity_int32/1/.
+cp libtriton_identity.so $DATADIR/custom_identity_int32/1/.
```
Comment (GuanLuo) [inline qa/L0_client_memory_growth/test.sh:57]: Same as comment in L0_batcher

**Review by GuanLuo — CHANGES_REQUESTED** [review]

File: Dockerfile.QA (hunk 1 above)
Comment (CoderHam) [reply Dockerfile.QA:213]: Sure thin

File: qa/L0_client_memory_growth/models/custom_identity_int32/config.pbtxt (hunk 2 above)
Comment (CoderHam) [reply qa/L0_client_memory_growth/models/custom_identity_int32/config.pbtxt:31]: Why not?

File: Dockerfile.QA [hunk 4]
```diff
@@ -165,51 +165,51 @@ RUN mkdir -p qa/custom_models/custom_sequence_int32/1 && \
     cp /tmp/tritonbuild/tritonserver/build/test-util/install/lib/libdyna_sequence.so \
         qa/custom_models/custom_dyna_sequence_int32/1/.
 
-RUN cp /tmp/tritonbuild/tritonserver/build/test-util/install/lib/libidentity.so \
+RUN cp /tmp/tritonbuild/install/backends/identity/libtriton_identity.so \
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply Dockerfile.QA:168]: Ah yes, I see

File: qa/L0_client_memory_growth/test.sh (hunk 3 above)
Comment (GuanLuo) [reply qa/L0_client_memory_growth/test.sh:57]: > And you don't need the library copy as the backend lookup will check the global location, where we place identity backend as part of the build.

File: build.py [hunk 5]
```diff
@@ -64,7 +64,7 @@
 # incorrectly load the other version of the openvino libraries.
 #
 TRITON_VERSION_MAP = {
-    '2.10.0dev': ('21.05dev', '21.03', '1.7.1', '2021.2.200', '2021.2.200')
+    '2.10.0dev': ('21.04dev_hemantj', '21.03', '1.7.1', '2021.2.200', '2021.2.200')
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply build.py:67]: Yup

File: qa/L0_batcher/test.sh [hunk 6]
```diff
@@ -408,7 +408,7 @@ if [[ $BACKENDS == *"custom"* ]]; then
     rm -fr ./custom_models && mkdir ./custom_models && \
         cp -r ../custom_models/custom_zero_1_float32 ./custom_models/. && \
         mkdir -p ./custom_models/custom_zero_1_float32/1 && \
-        cp ./libidentity.so ./custom_models/custom_zero_1_float32/1/libcustom.so
+        cp ./libtriton_identity.so ./custom_models/custom_zero_1_float32/1/libcustom.so
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply qa/L0_batcher/test.sh:411]: The model config was indeed updated for the new schema here: https://github.com/triton-inference-server/server/blob/hemantj-identity/qa/custom_models/custom_zero_1_float32/config.pbtxt
And the model does indeed load. I have removed the copy as you are right, it is not needed

File: Dockerfile.QA (hunk 4 above)
[the remark this answers is not in the document]
Comment (CoderHam) [reply Dockerfile.QA:168]: When I remove these and try running the test I get the following error:
```
model_repository_manager.cc:1451] failed to load model 'nop_TYPE_UINT64_-1': at least one version must be available under the version policy of model 'nop_TYPE_UINT64_-1'
```

Even thought the backend is loaded
```
+-------------+-----------------------------------------------------------------+-----------------------------+
| Backend     | Path                                                            | Config                      |
+-------------+-----------------------------------------------------------------+-----------------------------+
| identity    | /opt/tritonserver/backends/identity/libtriton_identity.so       | {}                          |
```

[the remark this answers is not in the document]
Comment (GuanLuo) [reply Dockerfile.QA:168]: You still need to create the version folder.. Can you double check that?

**Review by GuanLuo — CHANGES_REQUESTED** [review]

File: Dockerfile.QA (hunk 4 above)
[the remark this answers is not in the document]
Comment (CoderHam) [reply Dockerfile.QA:171]: Yes, removed the models. You were right

File: qa/L0_client_memory_growth/test.sh (hunk 3 above)
Comment (CoderHam) [reply qa/L0_client_memory_growth/test.sh:57]: Removed

File: Dockerfile.QA [hunk 7]
```diff
@@ -165,52 +165,50 @@ RUN mkdir -p qa/custom_models/custom_sequence_int32/1 && \
...
+    cp /tmp/tritonbuild/install/backends/identity/libtriton_identity.so \
         qa/L0_batcher/. && \
-    cp /tmp/tritonbuild/tritonserver/build/test-util/install/lib/libidentity.so \
+    cp /tmp/tritonbuild/install/backends/identity/libtriton_identity.so \
         qa/L0_client_timeout/. && \
-    cp /tmp/tritonbuild/tritonserver/build/test-util/install/lib/libidentity.so \
+    cp /tmp/tritonbuild/install/backends/identity/libtriton_identity.so \
         qa/L0_client_valgrind/. && \
-    cp /tmp/tritonbuild/tritonserver/build/test-util/install/lib/libidentity.so \
+    cp /tmp/tritonbuild/install/backends/identity/libtriton_identity.so \
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply Dockerfile.QA:213]: Removed

File: qa/L0_batcher/test.sh [hunk 8]
```diff
@@ -399,157 +399,6 @@ for i in \
     set -e
 done
 
-# Test that verify the 'preserve_ordering' option in dynamic batcher
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply qa/L0_batcher/test.sh:402]: Fixed

File: qa/L0_multi_server/test.sh [hunk 9]
```diff
@@ -64,6 +64,7 @@ for (( I=1; I<${THREAD_COUNT}+2; I++ )); do
         cp $DATADIR/${full}/config.pbtxt ${MODELSDIR}${I}/simple${I}/. && \
         (cd ${MODELSDIR}${I}/simple${I} && \
                 sed -i "s/^name:.*/name: \"simple${I}\"/" config.pbtxt && \
+                sed -i "s/^version_policy:.*/version_policy: { specific { versions: [1] }}/" config.pbtxt && \
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply qa/L0_multi_server/test.sh]: Removed. It is not necessary. The version policy was 1,3 but technically should have been 1. Not sure why it does not throw and error

File: qa/L0_lifecycle/test.sh [hunk 10]
```diff
@@ -51,1092 +51,1093 @@ rm -fr *.log
...
-    set -e
-
-    kill $SERVER_PID
-    wait $SERVER_PID
-
-    unset TRITONSERVER_USE_GRPC
-
-    LOG_IDX=$((LOG_IDX+1))
-done
+# SERVER_ARGS="--model-repository=/tmp/xyzx --strict-readiness=true \
```
Comment (GuanLuo) [inline qa/L0_lifecycle/test.sh:54]: Enable all lifecycle tests

File: qa/L0_model_config/test.sh [hunk 11]
```diff
@@ -45,7 +45,8 @@ source ../common/util.sh
 
 export CUDA_VISIBLE_DEVICES=0
 
-TRIALS="tensorflow_savedmodel tensorflow_graphdef tensorrt_plan onnxruntime_onnx pytorch_libtorch custom"
+# Removing custom backend since legacy custom backend is deprecated
```
Comment (GuanLuo) [inline qa/L0_model_config/test.sh:48]: Don't need this comment. No point of mentioning legacy backend after it is deprecated

**Review by GuanLuo — CHANGES_REQUESTED** [review]

File: qa/L0_multi_server/test.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (GuanLuo) [reply qa/L0_multi_server/test.sh]: The behavior is that error will only be logged when some of the specified versions are missing, load fail only if no version can be loaded

File: qa/L0_lifecycle/test.sh [hunk 12]
```diff
@@ -50,7 +50,7 @@ rm -fr *.log
 
 LOG_IDX=0
 
-# LifeCycleTest.test_parse_error_noexit_strict
+LifeCycleTest.test_parse_error_noexit_strict
```
Comment (GuanLuo) [inline qa/L0_lifecycle/test.sh:53]: Comment

**Review by GuanLuo — CHANGES_REQUESTED** [review]
You just need to fix the test script, otherwise the PR looks good to me.

**Review by GuanLuo — DISMISSED** [review]

**Comment by CoderHam** [comment]
@GuanLuo should I disable the trace test from the batcher since it looks like it is still broken?

**Comment by CoderHam** [comment]
L0_lifecycle failure will be resolved by https://github.com/triton-inference-server/backend/pull/23

**Comment by GuanLuo** [comment]
As I mentioned in the DM, master L0_batcher and L0_cmdline_trace passes so it is less likely that the trace is still broken.

L0_cmdline_trace shows that the trace is working for backend uses backend API. And even though the preserve ordering test in L0_batcher still uses the legacy custom backend, the code path for tracing should be identical to backends using backend APIs.

Can you elaborate on how you narrow down the cause back to trace?

File: build.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (CoderHam) [reply build.py]: Done

**Review by GuanLuo — APPROVED** [review]

**Pull request merged**